### PR TITLE
Inventory Rewrite - new simple-inventory plugin

### LINKED ISSUE #4215: /imports/included/inventory/server/init.js Not suitable for large stores.
## Steps to Reproduce
Please provide starting context, i.e. logged in as a user, configure a particular payment method.
1.  Create a store 
2.  Import over a million products
3. Without tuning the node vm, RC will OOM and crash

## Possible Solution
Include  a flag to build inventory as a background job instead of inside of "afterCoreInit".  The job should pluck id's by the 1000's or 10 of 1000's and perform bulk updates of the inventory collection to be performant.

```
import { Hooks, Logger } from "/server/api";
import { Products, Inventory } from "/lib/collections";
import { registerInventory } from "../methods/inventory";

// On first-time startup init the Inventory collection with entries for each product
Hooks.Events.add("afterCoreInit", () => {
  // If we already have any inventory record, skip
  const inventory = Inventory.find().count();
  if (!inventory) {
    const products = Products.find().fetch();
    for (const product of products) {
      Logger.debug(`Registering product ${product.title}`);
      registerInventory(product);
    }
  }
});

```
In the above code, the dangerous method is Products.find().count(), for most e-commerce stores, this wont cause an issue, for those with millions of Products this code isn't suitable.     Perhaps a flag to turn this off is a good idea.  A larger deployment will likely need a custom means of managing inventory

### LINKED ISSUE #4931: When cancelling a non-approved order with "Yes, Restock", inventoryInStock updates when it shouldn't
## Issue Description
There are two inventory tracking numbers:  `inventoryInStock` and `inventoryAvailableToSell`. When an order is placed, `inventoryAvailableToSell` should decrement.

When an operator processes this order, if the order is cancelled before payment is captured, AND the operator choose the option `Yes, restock`, then only the `inventoryAvailableToSell` number should update with the stock. Currently, both `inventoryInStock` and `inventoryAvailableToSell` update.

If the order payment has been captured, then both numbers should update.

## Steps to Reproduce
1. View inventory numbers
1. Create an order
1. Cancel the order before the payment is approved, and choose `Yes, restock` to restock inventory
1. See that both  `inventoryInStock` and `inventoryAvailableToSell` were replenished.

## Possible Solution

What seems to be happening is [this `orderIsApproved` check](https://github.com/reactioncommerce/reaction/blob/feat-kieckhafer-surcharges/imports/plugins/core/inventory/server/no-meteor/startup.js#L22) is now always returning `true` because `payment.status` has already been updated to be `cancelled` at this point, since this function is called in `afterOrderCancel`.

We could check for `payment.status !== completed` but we need to see how exactly that works with multiple payments. If one payment is completed, but a second isn’t, and the order is cancelled, does that refund the already processed payment and set it to `cancelled` too?

We could add a `beforeOrderCancel` appEvent at the beginning of [this function here](https://github.com/reactioncommerce/reaction/blob/develop/imports/plugins/core/orders/server/methods/cancelOrder.js#L21), but I'm not sure if that would serve any purpose aside for this specific issue.

### LINKED ISSUE #4932: Do not change inventoryInStock and inventoryAvailableToSell from null to 0
## Prerequisites
* [x] Are you running the latest version?
* [x] Are you able to consistently reproduce the issue?
* [x] Did you search the issue queue for existing issue? [Search issues](https://github.com/reactioncommerce/reaction/issues)

## Issue Description
In https://github.com/reactioncommerce/reaction/blob/develop/imports/plugins/core/catalog/server/no-meteor/utils/createCatalogProduct.js#L30-L31 we have `|| 0` for `inventoryInStock` and `inventoryAvailableToSell` everywhere. This would convert `null` to `0`, but these might be `null` if we are not tracking inventory, and converting to `0` would change the meaning. `null` means not tracked while `0` means tracked and unavailable.

## Possible Solution
Remove `|| 0` for `inventoryInStock` and `inventoryAvailableToSell` everywhere in that file. This is a simple task, but we will need to test carefully that this does not break anything in the storefront that might be expecting always to get a number and not properly handling `null` value.

### Versions
2.0.0 RC9

### LINKED ISSUE #5107: Move inventory data out of Product and into it's own collection
Having inventory on product causes problems for trying to update it as it requires that an entire product be published just to update inventory.

It should be its own standalone collection

### LINKED ISSUE #5123: Can't backorder variants if inventoryManagement is false
## Issue Description
When inventoryManagement and inventoryPolicy are disabled for a variant,  I am not allowed to order more than the available quantity.

I don't expect to have to turn on inventoryManagement to be able to allow backorders.

## Steps to Reproduce
Please provide starting context, i.e. logged in as a user, configure a particular payment method.
1. Create a product with variant
2. In the variant,  turn on allow Backorder ( just be sure, it will be greyed out in the next step)
3. Turn of Manage inventory
4. Publish
5. Try to order any amount higher than the quantity that variant has
6. Get error "Quantity ordered is more than available inventory for ..."

## Possible Solution
The issue happens during publish and is not related to the order process.
Catalog item does not have canBackorder = true as the check fails.

`buildOrderItem` fails here, as canBackorder is missing:
https://github.com/reactioncommerce/reaction/blob/6c8e8066e7cf36057f93bbf4f08e836676c44896/imports/plugins/core/orders/server/no-meteor/util/buildOrderItem.js#L29

`createCatalogProduct` is using a separate funtion to determine wether canBackorder should be `true`:
https://github.com/reactioncommerce/reaction/blob/6c8e8066e7cf36057f93bbf4f08e836676c44896/imports/plugins/core/catalog/server/no-meteor/utils/createCatalogProduct.js#L2

`canBackorder` will only return if inventoryManagement is `true`:
https://github.com/reactioncommerce/reaction/blob/6c8e8066e7cf36057f93bbf4f08e836676c44896/imports/plugins/core/catalog/server/no-meteor/utils/canBackorder.js#L9

**Why is this check necessary?**
**canBackorder should always be true when inventoryPolicy is false or not present**

**It should be just:**
``
export default function canBackorder(variants) {
  return variants.every((variant) => !variant.inventoryPolicy);
}
``

## Further thinking

Maybe that check should be the opposite of the original:
only return false if inventoryManagement and inventoryPolicy are both true.

But currently, the product admin UI allows to set "Manage inventory" and "Allow backorder" independently, altough I would expect "Manage inventory
" turned of to grey out "Allow backorder" aswell.

But Im not sure what usecases might be there, maybe this is on purpouse?

For example someone needs to limit the quantity per order, which would be possible with that current implementation were both values are decoupled. <-- **But this is still confusing and should, if necessary be implemented separately, as it is more of a trick using the current implementation.**

### Versions
v2.0.0-rc.11

### PR DESCRIPTION
**Pull request by aldeed**
Resolves #4215
Resolves #4931 
Resolves #4932 
Resolves #5107 
Resolves #5123   
Impact: **breaking**  
Type: **feature|performance**

### Overview
- Inventory code that was mixed up with Catalog and Product code is now properly isolated in inventory plugins.
- The "Inventory" collection is no longer created or available on `context.collections`. It was no longer used. It is not dropped, though, in case anyone is using it.
- Previously some inventory information was stored in the `Catalog` collection, on the `variants` and `options` objects. Now all of these fields are removed from the database except three top-level product bools that must be in the collection data for sorting and filtering purposes. This doesn't affect GraphQL, though, because those fields are still set at read time.
- The whole idea of auto-publishing inventory is mostly gone due to the above change, so the `AUTO_PUBLISH_INVENTORY_FIELDS` environment variable no longer affects anything.
- `buildOrderItem`, part of the process of placing an order, is updated to query inventory using the new API
- In the operator UI, the product table column for inventory flags is removed, and the variant column table for quantity is removed. These could be added back in the future if they are necessary, but it was extra complexity for the purposes of this initial rewrite.

### Core Inventory Plugin
There was already a core `inventory` plugin, but much of the inventory code was still scattered in other plugins. Now all inventory code is in the core `inventory` plugin, and plugins do not import or reference inventory code outside of the standard ways (context).

Here's what the core `inventory` plugin now does:
- Sets `product.isSoldOut`, `product.isLowQuantity`, and `product.isBackorder` booleans on a `CatalogItem` record when it is published. Then keeps these values correct whenever inventory information changes for a variant. We must store these fields in the DB rather than adding them at read time because we allow sorting and/or filtering by them.
- Adds inventory-related fields to `CatalogProductVariant` and `CartItem` at read time by registering transform functions that the `catalog` plugin calls.
- Adds `context.queries.inventoryForProductConfigurations`, and a convenience function for a single configuration, `context.queries.inventoryForProductConfiguration`, as the standard way for other plugins to retrieve inventory info.
    - These functions delegate to one or more other inventory plugins, such as the included `simple-inventory` plugin, for the actual data.
- Registers `isSoldOut`, `isLowQuantity`, and `isBackorder` as supported boolean filters for the `catalogItems` query.

### Simple Inventory Plugin
As mentioned above, the core `inventory` plugin requires one or more inventory plugins to do the actual inventory tracking for product configurations (variants).

In general, an inventory plugin needs to do the following:
- Register an `inventoryForProductConfigurations` type function that returns current inventory in stock and other related data for a list of product configurations.
- Optionally provide a way of editing the inventory in stock in the operator UI, by registering client components
- Optionally track order status changes to keep track of "reserved" inventory, which is inventory that is still technically in stock but should not be considered available to sell.

A new included `simple-inventory` plugin does all of the above in a manner similar to the way inventory worked before this PR. The UI has been improved a bit but is still in the same place. One major difference is that the actual inventory data is now stored in a new `SimpleInventory` collection rather than on `Products`, which makes it much easier to read and write. A migration populates `SimpleInventory` collection from data currently on `Products`, so the change should be transparent.

One UI improvement is regarding the "reserved" quantity for a variant. The UI now shows in stock, reserved, and available numbers, and there is a button that allows you to force the "reserved" number to update if it becomes incorrect. This plugin considers inventory to be reserved when it's part of an order that has been placed but is not yet approved.

### How to sync inventory quantities from an external system
**Option 1** If your external inventory system can be queried in real time, the simplest integration is to remove the `simple-inventory` plugin and add your own custom inventory plugin that registers an `inventoryForProductConfigurations` type function that returns current inventory in stock and other related data for a list of product configurations by calling through to your inventory management system.

**Option 2** If your external inventory system produces periodic exports but cannot be queried in real time, then you may want to keep the included `simple-inventory` plugin and create a script that reads the exported files and updates the `SimpleInventory` collection as necessary. If you do this in a custom plugin, then it's best to call `context.mutations.updateSimpleInventory` with `context.isInternalCall` set to `true`, and let that take care of properly updating the collection. If you can't do it in a custom plugin, then your external script should be written to look similar to the `updateSimpleInventory` mutation function in the `simple-inventory` plugin.

## Breaking changes
The GraphQL API is mostly unchanged. Inventory fields for `CatalogProductVariant` and `CartItem` are as they were. But internally, everything is different. So custom plugins that manage or use inventory are likely to need a rewrite.

The example web storefront will break unless you have the changes in https://github.com/reactioncommerce/reaction-next-starterkit/pull/527.

## Testing
This change touches a lot of areas, and although most have decent test coverage, we should carefully test everything that might be affected.

- Make sure there are no errors logged when migration 63 runs. Make sure inventory data looks the same on sellable variants before and after pulling this code and running the migration.
- Verify you can publish a product to the catalog with no errors
- Test all fields on the Inventory card of a sellable variant's edit page
- Manually set the `inventoryReserved` number to an incorrect number in the `SimpleInventory` collection and verify that it is fixed when clicking the "Recalculate reserved quantity" button
- On the storefront product grid, verify that the badge is correct.
    - Find a product and adjust its variant quantities such that you would expect "Sold Out" to appear for the product. Without needing to publish, you should be able to refresh the storefront page and see the "Sold Out" badge.
    - Do the same for "Low Quantity"
    - Do the same for "Backordered"
    - Do the same for no badge.
- With inventory management turned OFF for a variant, verify that you can add any quantity of that variant to your cart, and successfully place the order.
- With inventory management turned ON for a variant and "Allow backorder" turned OFF, verify that you cannot add more than the available quantity to your cart. There may be some edge case ways to add them to your cart but verify that placing the order fails.
- With inventory management turned ON for a variant and "Allow backorder" turned ON, verify that you can add more than the available quantity to your cart, and placing the order succeeds.
- Verify that filtering out sold out products when doing the `catalogItems` GraphQL query works.
- Reserved quantity testing:
    - After you place an order for a variant, if you look at the Inventory card for that variant in the operator UI, you should see the ordered quantity listed as reserved.
    - If you approve the fulfillment group, the reserved quantity should decrease and the in stock quantity should decrease.
    - If you cancel an item before approving, the reserved quantity should decrease and the in stock quantity should stay the same.
    - If you cancel an item after approving and choose to "return to stock", the reserved quantity should stay the same and the in stock quantity should increase.
    - If you cancel an item after approving and choose NOT to "return to stock", the reserved quantity should stay the same and the in stock quantity should stay the same.

### CODE REVIEW

**Comment by kieckhafer** [comment]
It seems the `SimpleInventory` is setting the ID of it's data to an `ObjectId`. @mikemurray and I came across this error earlier when investigating some Tag issues. Not sure if this is related to the issue where I can't update the fields or not. I don't know if this is an error or if this is something we are doing going forward, but we should make sure our `_id`'s are all the same Type.

Robo_3T_-_1_1

**Comment by aldeed** [comment]
@kieckhafer I pushed fixes for `_id` and the UI error you were getting. I added something late yesterday that started breaking the UI. I'll look at and address your other comments.

File: imports/plugins/included/simple-inventory/client/VariantInventoryForm.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                  }
+                });
+              }}
+            />
+          }
+          label={i18next.t("productVariant.isInventoryManagementEnabled")}
+        />
+        <Grid container spacing={8}>
+          <Grid item sm={6}>
+            <Components.TextField
```
[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/included/simple-inventory/client/VariantInventoryForm.js]: Yes I'd like to do this if there's time, or go back and do it after merge.

File: imports/plugins/core/versions/server/util/findAndConvertInBatchesNoMeteor.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import Logger from "@reactioncommerce/logger";
```
[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/core/versions/server/util/findAndConvertInBatchesNoMeteor.js]: Good idea

File: imports/plugins/core/inventory/server/no-meteor/schemas/schema.graphql [hunk 3]
```diff
@@ -8,3 +8,105 @@ extend enum CatalogBooleanFilterName {
...
+  inventoryAvailableToSell: Int
+
+  """
+  True if this item is currently sold out but allows backorders. A storefront UI may use this
+  to decide to show a "Backordered" indicator.
+  """
+  isBackorder: Boolean!
+
+  """
+  True it this item has a low quantity in stock. A storefront UI may use this
```
[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/core/inventory/server/no-meteor/schemas/schema.graphql]: Good idea

File: imports/plugins/core/inventory/server/no-meteor/schemas/schema.graphql [hunk 4]
```diff
@@ -8,3 +8,105 @@ extend enum CatalogBooleanFilterName {
...
+  to decide to show a "Low Quantity" indicator.
+  """
+  isLowQuantity: Boolean!
+
+  """
+  True if this item is currently sold out. A storefront UI may use this
+  to decide to show a "Sold Out" indicator when `isBackorder` is not also true.
```
[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/core/inventory/server/no-meteor/schemas/schema.graphql]: Good idea

File: imports/plugins/core/inventory/server/no-meteor/schemas/schema.graphql [hunk 5]
```diff
@@ -8,3 +8,105 @@ extend enum CatalogBooleanFilterName {
...
+  """
+  True for a purchasable variant if it has a low quantity in stock. For non-purchasable variants, this is
+  true if at least one purchasable child variant has a low quantity in stock. A storefront UI may use this
+  to decide to show a "Low Quantity" indicator.
+  """
+  isLowQuantity: Boolean!
+
+  """
+  True for a purchasable variant if it is sold out. For non-purchasable variants, this is
+  true if every purchasable child variant is sold out. A storefront UI may use this
```
[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/core/inventory/server/no-meteor/schemas/schema.graphql]: Good idea

File: imports/plugins/core/inventory/server/no-meteor/schemas/schema.graphql [hunk 6]
```diff
@@ -8,3 +8,105 @@ extend enum CatalogBooleanFilterName {
...
+
+  """
+  True for a purchasable variant if it is sold out but allows backorders. For non-purchasable variants, this is
+  true if every purchasable child variant is sold out but allows backorders. A storefront UI may use this
+  to decide to show a "Backordered" indicator.
+  """
+  isBackorder: Boolean!
+
+  """
+  True for a purchasable variant if it has a low quantity in stock. For non-purchasable variants, this is
```
[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/core/inventory/server/no-meteor/schemas/schema.graphql]: Good idea

File: imports/plugins/core/catalog/server/methods/catalog.js [hunk 7]
```diff
@@ -530,17 +412,11 @@ Meteor.methods({
     const isOption = ancestors.length > 1;
     if (isOption) {
       Object.assign(newVariant, {
-        title: `${parent.title} - Untitled option`
+        optionTitle: "Untitled",
```
[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/core/catalog/server/methods/catalog.js]: I don't remember why this ended up here, but it seems like it's correct this way.

**Comment by aldeed** [comment]
Remaining code comments have been addressed

**Comment by kkuzemka** [comment]
@kieckhafer - can you take another look at this in hopes that we can try and get it merged and over to SDI today?

**Comment by kieckhafer** [comment]
@kkuzemka will re-test first thing this morning.

**Comment by kieckhafer** [comment]
@aldeed All above mentioned issues related to the Orders UI seem to be fixed. Still seeing the issues we discussed with startup and packages loading.

**Comment by aldeed** [comment]
I fixed the issue with Package registry being blank on first run with blank DB. Migration 3 was clearing it so I commented out migration 3. :sweat_smile: I doubt anyone is upgrading from that far back, and it isn’t needed for a fresh install, so it seems like a safe way to fix.

**Pull request merged**